### PR TITLE
DAOS-4584 daos(8): Integrate obj_ctl into daos(8)

### PR DESCRIPTION
**Pull request by berserk-fury**
* Update memcheck-cart.supp suppress file.
* Merge the obj_ctl functionality into daos tool.

Signed-off-by: Martinez Montes, Jonathan <jonathan.martinez.montes@intel.com>

### CODE REVIEW

File: src/utils/obj_ctl.c
Comment (kccain) [inline src/utils/obj_ctl.c:557]: is this fprintf call redundant with the one above, line 507, if (argc < 3)? Also that printf has 2 string placeholders: %s %s daos|vos (filled with argv[0] and argv[1]) but this call has just one %s placeholder filled with argv[0].

**Review by kccain — CHANGES_REQUESTED** [review]
I looked at daos.c and obj_ctl.[ch] files more carefully than the others e.g., SConscript files.

**Review by jolivier23 — DISMISSED** [review]
change looks fine to me

File: src/utils/obj_ctl.c
Comment (berserk-fury) [reply src/utils/obj_ctl.c]: Done

**Review by jolivier23 — APPROVED** [review]

**Review by kccain — DISMISSED** [review]

**Review by bfaccini — APPROVED** [review]

**Comment by jolivier23** [comment]
Have you reviewed the test failures. There appear to be ones I haven't see on other patches

**Review by jolivier23 — CHANGES_REQUESTED** [review]
Looks like you lost a race on the daos.spec

**Comment by berserk-fury** [comment]
I ran the tests on Wolf using the following commands:

`$ POOL_SCM_SIZE=8 POOL_NVME_SIZE=0 daos_test -i`
`$ POOL_SCM_SIZE=8 POOL_NVME_SIZE=0 daos_test -r`

All of failed tests passed except IO31 which was skipped.

`[  SKIPPED ] IO31: fetch when all replicas unavailable`

I am wondering if I faced intermittent failures.
Anyway, I probably should wait for the tests to complete. The job queue is quite long

**Comment by berserk-fury** [comment]
It looks like the broken tests were not related to this PR, The latest report says:
"There are 0 new tests failing"

**Comment by ashleypittman** [comment]
There are a number of valgrind errors showing in Jenkins for this PR

https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-3440/8/valgrindResult/

**Comment by berserk-fury** [comment]
@ashleypittman I built today's master and ran a couple of commands using Valgrind. Valgrind reports similar errors for master and for my PR. I am wondering that those memory leaks might not belong to my PR :p

Full outputs are available on [Jira](https://jira.hpdd.intel.com/browse/DAOS-4584). Should I report those errors on a different ticket?

**Comment by berserk-fury** [comment]
@ashleypittman The valgrind errors are gone
https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-3440/13/valgrindResult/

File: src/cart/utils/memcheck-cart.supp
Comment (ashleypittman) [inline src/cart/utils/memcheck-cart.supp:127]: I think these would all be the same if you removed the match-leak-kinds field.

Regardless, we do at some point need to go through and audit/remove/fix all these issues as they all represent a bug, either in our code, or in OFI, or lower down the stack, or a missed close/free call in daos itself.

File: src/cart/utils/memcheck-cart.supp
Comment (ashleypittman) [inline src/cart/utils/memcheck-cart.supp:133]: Is there a rte_log_fini() function we should call?

At least one other person has also found this, so we should report upstream if there is no way to resolve it with the current spdk. https://tracker.ceph.com/issues/36033

File: src/cart/utils/memcheck-cart.supp
Comment (ashleypittman) [inline src/cart/utils/memcheck-cart.supp:157]: Are we using deprecated APIs?

**Review by ashleypittman — COMMENTED** [review]
I haven't check when this was last rebased but it would need to have the recent CaRT build/test changes included before merging because of the potential conflicts.

File: src/cart/utils/memcheck-cart.supp
Comment (berserk-fury) [reply src/cart/utils/memcheck-cart.supp:105]: The `rte_log_init()` funcion is called using a constructor here:
[eal_common_log.c](https://github.com/spdk/dpdk/blob/3fcb1dd395a870c64450f151a53338034b832906/lib/librte_eal/common/eal_common_log.c#L356)

I couldn’t find a `rte_log_fini()` or similar function. Can I file a ticket to investigate/address it?

File: src/cart/utils/memcheck-cart.supp
Comment (berserk-fury) [reply src/cart/utils/memcheck-cart.supp:129]: Yes, we are. The macro [SPDK_RPC_REGISTER](https://github.com/spdk/spdk/blob/b90630a465bf191b9bd3b0c2e4f2ba47b4044143/lib/rpc/rpc.c#L406) defines a constructor in [rpc.h](https://github.com/spdk/spdk/blob/b90630a465bf191b9bd3b0c2e4f2ba47b4044143/include/spdk/rpc.h#L130) that calls `spdk_rpc_register_alias_deprecated()`

File: src/cart/utils/memcheck-cart.supp
Comment (berserk-fury) [reply src/cart/utils/memcheck-cart.supp]: Done, I removed the math-leak-kins field.

My PR is up-to-date with latest master. It is on top of 058fbe5 (06/Oct/20)

File: src/cart/utils/memcheck-cart.supp
Comment (ashleypittman) [reply src/cart/utils/memcheck-cart.supp:105]: I think it's low priority, but we do need to check for these functions, and it would be good to file a upstream ticket.  This is low priority however, but carrying these suppressions does give us a maintenance burden.

File: src/cart/utils/memcheck-cart.supp
Comment (ashleypittman) [reply src/cart/utils/memcheck-cart.supp:129]: I'm not familiar with where we call these, and a quick "git grep" doesn't show anything.  Is there a ticket needed by us to update the API that we're using somewhere?

Comment (berserk-fury) [reply src/cart/utils/memcheck-cart.supp:129]: These memory leaks are isolated to the SPDK libraries. Moreover, these functions are called using [constructors](https://www.geeksforgeeks.org/__attribute__constructor-__attribute__destructor-syntaxes-c/). 
I reproduced the memory leaks without using DAOS code. Here is the minimal example:

`#include <stdio.h>

int main(int argc, char *argv[]) {
  printf("Hello SPDK\n");

  return 0;
}`

Then, I built it and linked it to the spdk libraries

`$ gcc -g -L/home/jonathan/repos/daos/install/prereq/dev/spdk/lib -lspdk_rpc -lspdk_env_dpdk -lrte_bus_pci -o hello_spdk my_program.c`

Finally, I called valgrind

`$ valgrind --fair-sched=yes --leak-check=full --show-leak-kinds=all ./hello_spdk`

I got the same memory leaks:

==12888== 96 bytes in 2 blocks are still reachable in loss record 4 of 6
==12888==    at 0x4C2C089: calloc (vg_replace_malloc.c:762)
==12888==    by 0x4E38DEE: spdk_rpc_register_alias_deprecated (rpc.c:300)
==12888==    by 0x400F9B2: _dl_init (in /usr/lib64/ld-2.17.so)
==12888==    by 0x4001179: ??? (in /usr/lib64/ld-2.17.so)

==12888== 249 bytes in 29 blocks are still reachable in loss record 5 of 6
==12888==    at 0x4C29F73: malloc (vg_replace_malloc.c:309)
==12888==    by 0x54E4AF9: strdup (in /usr/lib64/libc-2.17.so)
==12888==    by 0x6477DE2: rte_log_init (in /home/jonathan/repos/daos/build/external/dev/spdk/dpdk/build/lib/librte_eal.so.20.0)
==12888==    by 0x400F9B2: _dl_init (in /usr/lib64/ld-2.17.so)
==12888==    by 0x4001179: ??? (in /usr/lib64/ld-2.17.so)

[full output here](https://github.com/daos-stack/daos/files/5343882/minimal_example.txt)

As we can see, these functions are called automatically. According to [Valgrind documentation](https://www.valgrind.org/docs/manual/faq.html), the `still reachable` status is not necessary a true memory leak. I am wondering if it is a bug?

File: src/cart/utils/memcheck-cart.supp
Comment (berserk-fury) [reply src/cart/utils/memcheck-cart.supp:105]: See below comment for full details. I am wondering if it is bug that we should report to the SPDK team.

Valgrind documentation says:
"still reachable" means your program is probably ok -- it didn't free some memory it could have.

File: src/cart/utils/memcheck-cart.supp
Comment (ashleypittman) [reply src/cart/utils/memcheck-cart.supp:129]: Ok, so it shoulds like this is something that SPDK needs to or is working on.  What I wanted to know was that there wasn't any action required on our side to update to latest interfaces.

File: src/cart/utils/memcheck-cart.supp
Comment (ashleypittman) [reply src/cart/utils/memcheck-cart.supp:105]: "still reachable" is one of the types of memory leak we check for, currently there are zero still reachable regions in the results of our testing, and the threshold for failure is also zero, although we do suppress some.  If you're re-spinning this PR then can you add a comment that we've checked for a rte_log_fini() function and there isn't one, and it'll save time when it comes to reviewing the contents of the suppression file in future.

**Comment by berserk-fury** [comment]
@ashleypittman May I ask for your approval?

File: src/common/tests/common_test.h
Comment (ashleypittman) [inline src/common/tests/common_test.h:2]: Technically doesn't seem required.

File: src/utils/obj_ctl.c
Comment (ashleypittman) [inline src/utils/obj_ctl.c:514]: Is this change intentional?

**Review by ashleypittman — APPROVED** [review]
No objection from me.

File: src/utils/obj_ctl.c
Comment (berserk-fury) [reply src/utils/obj_ctl.c:514]: Yes, it is. Without it, I got this error:

`$ daos shell daos`
ERROR: dmg: pool create failed: server: code = 655 description = "requested NVMe capacity (4.0 GiB) is too small (min 8.0 GiB)"
ERROR: dmg: server: code = 655 resolution = "NVMe capacity should be larger than 8.0 GiB"

I got a similar error on master eac0d15 (10/Oct) while running the original tool:

`$ obj_ctl daos`
ERROR: dmg: pool create failed: server: code = 655 description = "requested NVMe capacity (4.0 GiB) is too small (min 8.0 GiB)"
ERROR: dmg: server: code = 655 resolution = "NVMe capacity should be larger than 8.0 GiB"

File: src/common/tests/common_test.h
Comment (jolivier23) [reply src/common/tests/common_test.h:2]: Probably happened due to a merge commit where someone didn't update it.   This happens to me quite a lot but hopefully less if people are installing the hook.

**Review by jolivier23 — APPROVED** [review]

File: src/utils/obj_ctl.c
Comment (mjmac) [reply src/utils/obj_ctl.c:514]: For the record, this is because of the cluster size defined in https://github.com/daos-stack/daos/blob/master/src/bio/bio_xstream.c#L46

Before, you could request less than 1.0GiB of NVMe per target, but you were still getting at least 1.0GiB. Now the tool lets you know.

**Review by jolivier23 — CHANGES_REQUESTED** [review]
sorry, one more update needed to resolve conflict with 1.1.1 spec change.

**Comment by brianjmurrell** [comment]
> sorry, one more update needed to resolve conflict with 1.1.1 spec change.

Which will also conflict with #3347 already in the queue.  I have not really figured out how to resolve this kind of in-queue conflict other than to wait for it to land and then rebase/merge.  :-(

**Review by jolivier23 — APPROVED** [review]

**Review by ashleypittman — APPROVED** [review]

**Review by bfaccini — APPROVED** [review]

**Pull request merged**
